### ISSUE TITLE
clang-format may treat aggregate initialization of a pointer type inside an if condition as multiplication

### ISSUE THREAD

**Issue by MarcoFalke** [llvm/llvm-project#63846]
Steps to reproduce with `clang-format version 16.0.6`:

* Output of `clang-format /tmp/a.cpp`:

```cpp
#define MYI int
int main() {
  if (2 * 3)
    ;
  if (int *p{nullptr})
    ;
  if (MYI * p{nullptr}) // wrong, should be "*p"
    ;
  if (MYI *p = nullptr)
    ;
}

**Comment by rymiel** [comment]
Note that this a fix for this was implemented in https://github.com/llvm/llvm-project/commit/70de684d44135b4025d92b2b36ad387cf5ab8b5a but was reverted in https://github.com/llvm/llvm-project/commit/104cd749f5cca609a79303c0dad22bc041b5448a due to regressions elsewhere

**Comment by owenca** [comment]
A new option `TypeNames` was added in 5c106f7b947e. Adding `MYI` to the list gives the expected format:
```
$ clang-format -style='{TypeNames: [MYI]}' a.cpp
#define MYI int
int main() {
  if (2 * 3)
    ;
  if (int *p{nullptr})
    ;
  if (MYI *p{nullptr})
    ;
  if (MYI *p = nullptr)
    ;
}
```

**Comment by MichaelSteffens** [comment]
The scaling concern raised here is addressed in #210182: keep the `TypeNames` whitelist
in a per-directory `.clang-format` using `BasedOnStyle: InheritParentConfig`, confined to
the code that names those types, instead of a growing global list. #210182 also asks for
a `PointerAlignment` → `TypeNames` cross-reference so users hitting this symptom find the
option. Cross-linking as evidence of the recurring pattern.
